lib/workers/repository/process/lookup/utils.ts

```diff
@@ -0,0 +1,91 @@
+import is from '@sindresorhus/is';
+
+import { getRangeStrategy } from '../../../../modules/manager';
+import type { LookupUpdate } from '../../../../modules/manager/types';
+import * as allVersioning from '../../../../modules/versioning';
+import * as template from '../../../../util/template';
+import type { LookupUpdateConfig } from './types';
+
+export function addReplacementUpdateIfValid(
+  updates: LookupUpdate[],
+  config: LookupUpdateConfig
+): void {
+  const replacementNewName = determineNewReplacementName(config);
+  const replacementNewValue = determineNewReplacementValue(config);
+
+  if (
+    config.packageName !== replacementNewName ||
+    config.currentValue !== replacementNewValue
+  ) {
+    updates.push({
+      updateType: 'replacement',
+      newName: replacementNewName,
+      newValue: replacementNewValue!,
+    });
+  }
+}
+
+export function isReplacementNameRulesConfigured(
+  config: LookupUpdateConfig
+): boolean {
+  return (
+    is.nonEmptyString(config.replacementName) ||
+    is.nonEmptyString(config.replacementNameTemplate)
+  );
+}
+
+export function isReplacementRulesConfigured(
+  config: LookupUpdateConfig
+): boolean {
+  return (
+    isReplacementNameRulesConfigured(config) ||
+    is.nonEmptyString(config.replacementVersion)
+  );
+}
+
+export function determineNewReplacementName(
+  config: LookupUpdateConfig
+): string {
+  // TODO - Discuss in PR.  What fields should be supported within replacementNameTemplate (all or a subset)
+  // TODO - Discuss in PR.  Where should packageNameWithoutRegistry be set?
```

Comment:
Where is the best place to be setting this new template field?

Reply:
It has to be added to the allowed fields. https://github.com/renovatebot/renovate/blob/main/lib/util/template/index.ts#L70

Not sure where the best place would be. Probably in the datasource as the registryUrl is parsed their from the packageName field. @viceice WDYT?

Reply:
not sure either. but we should use a more replacement related capture group name. because the replacement can also be a simple npm prefix. so not always related to registry

Reply:
I don't really know how to move this forward tbh... 

i cannot see how a capture group would be used to add a new registry, vs swapping a registry....

Reply:
This is most likely a "big deal" to achieve.... but throwing it out there....

For the docker datasource, is there a reason both `depName` and `packageName` are the same?

According to https://github.com/renovatebot/renovate/blob/main/lib/util/template/index.ts#L70, `packageName` is the full name.  What if `depName` was `packageName` without the registry?  Then we wouldn't need another template name that is too specific